test/xds/xds_client_ignore_resource_deletion_test.go

```diff
@@ -0,0 +1,474 @@
+// set in "server_features" field. This subtest verifies that the resource was
+// not deleted by the xDSClient when a resource is missing the xDS response and
+// RPCs continue to succeed.
+func testResourceDeletionIgnored(t *testing.T, resources e2e.UpdateOptions, u func(r *e2e.UpdateOptions)) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	t.Cleanup(cancel)
+	ms := startManagementServer(t)
+	bootstrapContent := generateBootstrapContents(t, ms.Address, true)
+
+	// Update the management server with initial resources setup.
+	if err := ms.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	cResolver := generateCustomResolver(t, bootstrapContent)
+	cc, err := grpc.Dial(fmt.Sprintf("xds:///%s", serviceName), grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(cResolver))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	t.Cleanup(func() { cc.Close() })
+
+	verifyRPCtoAllEndpoints(t, cc)
+
+	// Mutate resource and update on the server.
+	u(&resources)
+	if err := ms.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Make RPCs for every 50ms for the next 500ms.
+	timer := time.NewTimer(500 * time.Millisecond)
+	ticker := time.NewTicker(50 * time.Millisecond)
+	t.Cleanup(ticker.Stop)
+	for {
+		verifyRPCtoAllEndpoints(t, cc)
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			return
+		case <-ticker.C:
+		}
+	}
+}
+
+// This subtest tests the scenario where the bootstrap config has "ignore_resource_deletion"
+// not set in "server_features" field. This subtest verifies that the resource was
+// deleted by the xDSClient when a resource is missing the xDS response and subsequent
+// RPCs fail.
+func testResourceDeletionNotIgnored(t *testing.T, resources e2e.UpdateOptions, u func(r *e2e.UpdateOptions)) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout*1000)
+	t.Cleanup(cancel)
+	ms := startManagementServer(t)
+	bootstrapContent := generateBootstrapContents(t, ms.Address, false)
+
+	// Update the management server with initial resources setup.
+	if err := ms.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	cResolver := generateCustomResolver(t, bootstrapContent)
+	cc, err := grpc.Dial(fmt.Sprintf("xds:///%s", serviceName), grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(cResolver))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	t.Cleanup(func() { cc.Close() })
+
+	verifyRPCtoAllEndpoints(t, cc)
+
+	// Mutate resource and update on the server.
+	u(&resources)
+	if err := ms.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Spin up go routines to verify RPCs fail after the update.
+	client := testServicepb.NewTestServiceClient(cc)
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		for ctx.Err() == nil {
+			if _, err := client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+				wg.Done()
```

Comment:
Use `defer wg.Done()` at the top of the goroutine. That is a common pattern that we use all the time.